program-test/src/programs.rs

```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
-                })
-                .unwrap();
-                program_data.extend_from_slice(elf);
-                accounts.push((
-                    programdata_address,
-                    AccountSharedData::from(Account {
-                        lamports: rent.minimum_balance(program_data.len()).max(1),
-                        data: program_data,
-                        owner: *loader_id,
-                        executable: false,
-                        rent_epoch: 0,
-                    }),
-                ));
-                bincode::serialize(&UpgradeableLoaderState::Program {
-                    programdata_address,
-                })
-                .unwrap()
-            } else {
-                elf.to_vec()
-            };
-            accounts.push((
-                *program_id,
-                AccountSharedData::from(Account {
-                    lamports: rent.minimum_balance(data.len()).max(1),
-                    data,
-                    owner: *loader_id,
-                    executable: true,
-                    rent_epoch: 0,
-                }),
-            ));
-            accounts.into_iter()
+        .flat_map(|info| {
+            let accounts = program_accounts(rent, info);
+            Vec::<_>::from(accounts).into_iter()
         })
         .collect()
 }
+
+fn program_accounts(rent: &Rent, info: &ProgramInfo) -> ProgramAccounts {
+    let ProgramInfo {
+        program_id,
+        loader_id,
+        elf,
+    } = info;
+
+    match *loader_id {
+        bpf_loader::ID => {
+            let data = elf.to_vec();
+
+            let account = AccountSharedData::from(Account {
+                lamports: rent.minimum_balance(data.len()).max(1),
+                data,
+                owner: *loader_id,
+                executable: true,
+                rent_epoch: 0,
+            });
+
+            ProgramAccounts::NonUpgradable {
+                id: *program_id,
+                account,
+            }
+        }
+        bpf_loader_upgradeable::ID => {
+            let (main_account_data, program_data_id, program_data_account) =
+                upgradable_program_account(rent, info, 0, Some(Pubkey::default()));
+
+            let main_account = AccountSharedData::from(Account {
+                lamports: rent.minimum_balance(main_account_data.len()).max(1),
+                data: main_account_data,
+                owner: *loader_id,
+                executable: true,
+                rent_epoch: 0,
+            });
+
+            ProgramAccounts::Upgradable {
+                main_id: *program_id,
+                main_account,
+                program_data_id,
+                program_data_account,
+            }
+        }
+        unexpected => panic!("Unexpected program loader: {unexpected}"),
+    }
+}
+
+fn upgradable_program_account(
+    rent: &Rent,
+    ProgramInfo {
+        program_id,
+        loader_id,
+        elf,
+    }: &ProgramInfo,
+    last_modified_slot: u64,
```

Comment:
Looks like you're always setting this to `0`, so why not just set it to `0` within the helper, and don't pass it?
Maybe same with `upgrade_authority_address`?

Reply:
I think it is bad if a function has a generic name, but then is actually specialized to a certain particular use case.
Making these arguments explicit moves the specialization into the caller, where it is slightly more obvious.
Not perfect, but the caller is already specialized, so, at least, this way all the specialization happens in a single function.
And it is the API of the crate, an entry point - I think, a better place for specialization to be.
Compared to a lower level helper function.

Reply:
These are both private helpers, though. As per your argument, it would make more sense to provide the specialization on the module's entrypoint(s): `spl_program` and `spl_programs`. Considering their use case(s) that seems like overkill.

Unless of course you plan to export these helpers in the later commit you've mentioned, wherein you'll need to use those inputs?